### PR TITLE
fixes #3616: make sure there is always one time that we test the sock…

### LINKED ISSUE #3616: Spurious acme coverage failures
I've seen this a couple times now. We should fix this.

```
Name                           Stmts   Miss  Cover   Missing
------------------------------------------------------------
acme.py                            0      0   100%
acme/challenges.py               194      0   100%
acme/challenges_test.py          282      0   100%
acme/client.py                   233      0   100%
acme/client_test.py              347      0   100%
acme/crypto_util.py               92      2    98%   134-135
acme/crypto_util_test.py          79      0   100%
acme/dns_resolver.py              16      0   100%
acme/dns_resolver_test.py         36      0   100%
acme/errors.py                    29      0   100%
acme/errors_test.py               28      0   100%
acme/fields.py                    32      0   100%
acme/fields_test.py               41      0   100%
acme/jose.py                       8      0   100%
acme/jose/b64.py                  15      0   100%
acme/jose/b64_test.py             38      0   100%
acme/jose/errors.py               12      0   100%
acme/jose/errors_test.py           8      0   100%
acme/jose/interfaces.py           39      0   100%
acme/jose/interfaces_test.py      71      0   100%
acme/jose/json_util.py           170      0   100%
acme/jose/json_util_test.py      214      0   100%
acme/jose/jwa.py                 105      0   100%
acme/jose/jwa_test.py             58      0   100%
acme/jose/jwk.py                 114      0   100%
acme/jose/jwk_test.py            101      0   100%
acme/jose/jws.py                 205      0   100%
acme/jose/jws_test.py            146      0   100%
acme/jose/util.py                114      0   100%
acme/jose/util_test.py           128      0   100%
acme/jws.py                       17      0   100%
acme/jws_test.py                  27      0   100%
acme/messages.py                 174      0   100%
acme/messages_test.py            193      0   100%
acme/standalone.py                94      0   100%
acme/standalone_test.py           93      3    97%   145-147
acme/test_util.py                 33      0   100%
acme/util.py                      11      0   100%
acme/util_test.py                 17      0   100%
------------------------------------------------------------
TOTAL                           3614      5    99%
nose.plugins.cover: ERROR: TOTAL Coverage did not reach minimum required: 100%
```

In case line number changes, here's the relevant lines with some extra padding around them for additional context:

```
$ sed -n 129,140p crypto_util.py
        'source_address': source_address}

    try:
        # pylint: disable=star-args
        sock = socket.create_connection((host, port), **socket_kwargs)
    except socket.error as error:
        raise errors.Error(error)

    with contextlib.closing(sock) as client:
        client_ssl = OpenSSL.SSL.Connection(context, client)
        client_ssl.set_connect_state()
        client_ssl.set_tlsext_host_name(name)  # pyOpenSSL>=0.13
```

```
$ sed -n 140,152p standalone_test.py
        while max_attempts:
            max_attempts -= 1
            try:
                cert = crypto_util.probe_sni(
                    b'localhost', b'0.0.0.0', self.port)
            except errors.Error:
                self.assertTrue(max_attempts > 0, "Timeout!")
                time.sleep(1)  # wait until thread starts
            else:
                self.assertEqual(jose.ComparableX509(cert),
                                 test_util.load_comparable_cert('cert.pem'))
                break

```

### PR DESCRIPTION
**Pull request by schuyler1d**
…et error when starting a server

### CODE REVIEW

**Comment by schuyler1d** [comment]
ok, i think I addressed that issue by doing neither 1. or 2. but just fixing the fact that I didn't understand the original try...except...else construction.

**Comment by schuyler1d** [comment]
@bmw please re-review

File: acme/acme/standalone_test.py [hunk 1]
```diff
@@ -146,19 +145,22 @@ def tearDown(self):
...
-            max_attempts -= 1
+        for attempt in range(max_attempts):
             try:
                 cert = crypto_util.probe_sni(
                     b'localhost', b'0.0.0.0', self.port)
-            except errors.Error:
-                self.assertTrue(max_attempts > 0, "Timeout!")
-                time.sleep(1)  # wait until thread starts
-            else:
                 self.assertEqual(jose.ComparableX509(cert),
```
Comment (bmw) [inline acme/acme/standalone_test.py]: I think it'd be good to keep this in the `else:` block as it minimizes the amount of code in the `try` block which is largely the purpose of the `try`, `except`, `else` pattern in Python.

File: acme/acme/standalone_test.py [hunk 2]
```diff
@@ -146,19 +145,22 @@ def tearDown(self):
...
                                  test_util.load_comparable_cert(
                                      'rsa2048_cert.pem'))
                 break
+            except errors.Error:
+                self.assertTrue(max_attempts > 0, "Timeout!")
```
Comment (bmw) [inline acme/acme/standalone_test.py]: Now that we never change `max_attempts`, this assertion never fires. Instead, we should change it to something like:
```
self.assertTrue(attempt + 1 < max_attempts, "Timeout!")
```
which will achieve the same effect.

**Comment by schuyler1d** [comment]
@bmw ok, fixed -- sorry about the delay -- I didn't see your comment until yesterday

**Comment by SwartzCr** [comment]
woo hoo :tada:

**Pull request merged**